### PR TITLE
Further enhancements to core/test-extension and extension-authors-guide

### PR DESCRIPTION
**Pull request by starksm64**
The PR implements more of the testing outlined in #780 and adds a new Sample Test Extension section to the extension-authors-guide that illustrates several common extension tasks. A TOC was added and the header formatting was corrected to make the document more navigable.

### CODE REVIEW

File: core/test-extension/runtime/pom.xml [hunk 1]
```diff
@@ -53,6 +53,15 @@
...
         </dependency>
+        <!-- Don't rely on JDK JAXB so code compiles and runs under both JDK8/11 -->
+        <dependency>
+            <groupId>javax.xml.bind</groupId>
+            <artifactId>jaxb-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.glassfish.jaxb</groupId>
+            <artifactId>jaxb-runtime</artifactId>
+        </dependency>
```
Comment (gsmet) [inline core/test-extension/runtime/pom.xml]: I think it should be done in the `quarkus-jaxb` extension and you should depend on it.

File: core/test-extension/runtime/src/main/java/io/quarkus/extest/runtime/subst/KeyProxy.java [hunk 2]
```diff
@@ -1,5 +1,8 @@
 package io.quarkus.extest.runtime.subst;
 
+/**
+ *
+ */
```
Comment (gsmet) [inline core/test-extension/runtime/src/main/java/io/quarkus/extest/runtime/subst/KeyProxy.java]: Hmmm? Either we should have some javadoc or not.

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 3]
```diff
@@ -1,9 +1,10 @@
-include::./attributes.adoc[]
 = {project-name} - Writing Your Own Extension
-
 :numbered:
 :sectnums:
 :sectnumlevels: 4
+:toc:
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: We will need to check how it looks like on the website.

File: core/test-extension/deployment/src/test/java/io/quarkus/extest/ConfiguredBean.java [hunk 4]
```diff
@@ -1,20 +1,19 @@
 package io.quarkus.extest;
 
-import javax.enterprise.context.ApplicationScoped;
 import javax.enterprise.event.Observes;
-import javax.inject.Inject;
+import javax.enterprise.inject.spi.CDI;
```
Comment (gsmet) [inline core/test-extension/deployment/src/test/java/io/quarkus/extest/ConfiguredBean.java]: I don't see this import used (just checking on GH so I might be missing something)

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 5]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
 ./mvnw clean install -Dquarkus.debug.generated-classes-dir=./target/app-generated-classes
 ----
+
+
+== Sample Test Extension
+We have an extension that is used to test for regressions in the extension processing. It is located in {quarkus-tree-url}/core/test-extension directory. In this section we touch on some of the tasks an extension
+author will typically need to perform using the test-extension code to illustrate how the task could be done.
+
+=== Features and Capabilities
+When you startup a {project-name} instance, you will see a line like the one highlighted in this example:
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: s/startup/start/ ?

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 6]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+
+.TestProcessor#featureBuildItem()
+[source,java]
+----
+    /**
+     * Register a extension capability and feature
+     * 
+     * @return test-extension feature build item
+     */
+    @BuildStep(providesCapabilities = "io.quarkus.test-extension")
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: Not sure it helps having the capability defined here.

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 7]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+    FeatureBuildItem featureBuildItem() {
+        return new FeatureBuildItem("test-extension");
+    }
+----
+
+The feature name can be used to select the extension when creating a project using the {project-name} maven plugin as shown in this example taken from the Writing JSON REST Services guide where the "resteasy-jsonb" feature is referenced:
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: AFAIK, that is not true. The feature name is a totally different beast and is only used for display.

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 8]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+/**
+ * A sample bean
+ */
+@TestAnnotation <2>
+public class ConfiguredBean implements IConfigConsumer {
+
+...
+----
+<1> Register the annotation class and CDI default scope using the Jandex DotName class.
+<2> ConfiguredBean will be processed by the CDI layer the same as a bean annotated with the CDI standard @ApplicationScoped.
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: Some backticks to enclose the classnames would be nice.

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 9]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+
+=== Parsing Config to Objects
+One of the main things an extension is likely to do is completely separate the configuration phase of behavior from the runtime phase. Frameworks often do parsing/load of configuration on startup that can be done during build time to both reduce the runtime dependencies on frameworks like xml parsers as well as reducing the startup time the parsing incurs.
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: So I'm not a native speaker so it's worth what it's worth but I find these 2 sentences hard to parse, especially the second one. And I'm not sure I see what Quarkus brings here.

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 10]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+            log.infof("Have XmlConfig, loading");
+            XmlConfig config = (XmlConfig) unmarshaller.unmarshal(is); <2>
+...
+        }
+        return serviceBuildItem;
+    }
+
+----
+<1> Look for a config.xml classpath resource
+<2> If found, parse using JAXB context for `XmlConfig.class`
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: So while having an example of an XML file is nice, I think a lot of people use YAML nowadays.

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 11]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+
+[TIP]
+====
+If you look at the XmlConfig code you will see that it does carry around the JAXB annotations. If you don't want these in the runtime image, you could clone the XmlConfig instance into some POJO object graph and then replace XmlConfig with the POJO class. We will do this in <<Replacing Classes in the Native Image>>.
+====
+
+=== Scanning Deployments Using Jandex
+If your extension defines annotations or interfaces that mark beans needing to be processed, you can locate these beans using the Jandex API, a Java annotation indexer and offline reflection library. The following `TestProcessor#scanForBeans` method showns how to find the beans annotated with our `@TestAnnotation` tha also implement the `IConfigConsumer` interface:
+
+.x
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: Looks like you're missing a title.

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 12]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+
+If for some reason you need to parse the config and use it in other build steps in an extension processor, you would need to create an `XmlConfigBuildItem` to pass the parsed XmlConfig instance around. 
+
+[TIP]
+====
+If you look at the XmlConfig code you will see that it does carry around the JAXB annotations. If you don't want these in the runtime image, you could clone the XmlConfig instance into some POJO object graph and then replace XmlConfig with the POJO class. We will do this in <<Replacing Classes in the Native Image>>.
+====
+
+=== Scanning Deployments Using Jandex
+If your extension defines annotations or interfaces that mark beans needing to be processed, you can locate these beans using the Jandex API, a Java annotation indexer and offline reflection library. The following `TestProcessor#scanForBeans` method showns how to find the beans annotated with our `@TestAnnotation` tha also implement the `IConfigConsumer` interface:
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: s/showns/shows/
s/tha also/that also/

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 13]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+            } catch (ClassNotFoundException e) {
+                log.warn("Failed to load bean class", e);
+            }
+        }
+    }
+----
+<1> Depend on a `BeanArchiveIndexBuildItem` to have the build step be run after the deployment has been indexed.
+<2> Retrieve the index.
+<3> Find all beans annotated with `@TestAnnotation`.
+<4> Determine which of these beans also has the `IConfigConsumer` interface.
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: s/also has/also have/ ?

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 14]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+<5> Save the bean class in a `TestBeanBuildItem` for use in a latter RUNTIME_INIT build step that will interact with the bean instances.
+
+=== Interacting With Extension Beans
+You can use the `io.quarkus.arc.runtime.BeanContainer` interface to interace with your extension beans. The following `configureBeans` methods illustrate interacting with the beans scanned for in the previous section:
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: s/interace/interact/

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 15]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+    public void configureBeans(BeanContainer beanContainer, Class<IConfigConsumer> beanClass,
+            TestBuildAndRunTimeConfig buildTimeConfig,
+            TestRunTimeConfig runTimeConfig) {
+        log.infof("Begin BeanContainerListener callback\n");
+        IConfigConsumer instance = beanContainer.instance(beanClass); <4>
+        instance.loadConfig(buildTimeConfig, runTimeConfig); <5>
+        log.infof("configureBeans, instance=%s\n", instance);
+    }
+----
+<1> Consume the `TestBeanBuildItem`s produced from the scanning build step.
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: I wonder if you need double backticks here as you have a `s` just after.

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 16]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+        this.service = service;
+    }
+
+    public RuntimeValue<RuntimeXmlConfigService> getService() {
+        return service;
+    }
+}
+----
+<1> Call into the runtime template to record the creation of the service.
+<2> Using the parsed `XmlConfig` instance, create a instance of `RuntimeXmlConfigService` and wrap it in a `RuntimeValue`. Use a `RuntimeValue` wrapper for non-interface objects that are non-proxiable.
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: s/a instance/an instance/

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 17]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+            throws IOException {
+        RuntimeXmlConfigService service = runtimeValue.getValue();
+        service.startService(); <4>
+        shutdownContext.addShutdownTask(service::stopService); <5>
+    }
+----
+<1> We consume a ShutdownContextBuildItem to register the service shutdown.
+<2> We consume the previously initialized service captured in `RuntimeServiceBuildItem`.
+<3> Call the runtime template to record the service start invocation.
+<4> Runtime template retrieves the service instance reference and calls it's `startService` method.
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: s/it's/its/

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 18]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+<5> Runtime template registers an invocation of the service instance `stopService` method with the {project-name} `ShutdownContext`.
+
+The code for the `RuntimeXmlConfigService` can be viewed here:
+{quarkus-blob-url}/core/test-extension/runtime/src/main/java/io/quarkus/extest/runtime/RuntimeXmlConfigService.java[RuntimeXmlConfigService.java]
+
+The testcase for validating that the `RuntimeXmlConfigService` has started can be found in the `testRuntimeXmlConfigService` test of `ConfiguredBeanTest` and `NativeImageIT`.
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: s/testcase/test case/ maybe?

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 19]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+
+=== Register Resources for Use in Native Image
+Not all configuration or resources can be consumed at build time. If you have classpath resources that the runtime needs to access, you need to inform the build phase that these resource need to be copied into the native image. This is done by producing one or more `SubstrateResourceBuildItem` or `SubstrateResourceBundleBuildItem` in the case of resource bundles. Examples of this are shown in this sample `registerNativeImageReources` build step:
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: s/these resource/these resources/

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 20]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+        resourceBundle.produce(new SubstrateResourceBuildItem("javax.xml.bind.Messages")); <3>
+    }
+}
+----
+<1> Indicate that the /security/runtime.keys classpath resource should be copied into native image.
+<2> Indicate that the `io.quarkus.SomeService` ServiceLoader resource should be copied into native image.
+<3> Indicate that the "javax.xml.bind.Messages" resource bundle should be copied into native image.
+
+=== Object Substitution
+Objects created during the build phase that are passed into the runtime need to have a in order to be created and configured at startup of the runtime from the build time state. If an object does not have a default constructor you will see an error similar to the following during generation of the augmented artifacts:
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: `have a in order` <- there's a word missing, probably `default constructor`

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 21]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+            dsaPublicKey = (DSAPublicKey) kf.generatePublic(publicKeySpec);
+
+        } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
+            e.printStackTrace();
+        }
+        return dsaPublicKey;
+    }
+}
+----
+<1> The serialize method takes the object without a default ctor and creates a `KeyProxy` that contains the information neccessary to recreate the `DSAPublicKey`.
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: s/ctor/constructor/ maybe?
s/neccessary/necessary/

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 22]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+<2> The deserialize uses the `KeyProxy` to recreated the `DSAPublicKey` from it's encoded form using the key factory.
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: s/The deserialize/The deserialize method/
s/recreated/recreate/
s/it's/its/

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 23]
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+        ObjectSubstitutionBuildItem keysub = new ObjectSubstitutionBuildItem(holder);
+        substitutions.produce(keysub);
+
+        log.infof("loadDSAPublicKey run");
+        return new PublicKeyBuildItem(publicKey);
+    }
+----
+
+=== Replacing Classes in the Native Image
+The Graal SDK supports subsitutions of classes in the native image. An example of how one could replace the `XmlConfig/XmlData` classes with versions that have no JAXB annotation dependencies is shown in these example classes:
```
Comment (gsmet) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: s/subsitutions/substitutions/

**Review by gsmet — DISMISSED** [review]
Nice addition. I added some comments inline.

File: core/test-extension/runtime/pom.xml (hunk 1 above)
Comment (starksm64) [reply core/test-extension/runtime/pom.xml]: I would be concerned the core module is depending on too many modules outside of the core?

File: docs/src/main/asciidoc/extension-authors-guide.adoc (hunk 7 above)
Comment (mkouba) [reply docs/src/main/asciidoc/extension-authors-guide.adoc]: I think so too - names and labels used by CLI are currently hard-coded in https://github.com/quarkusio/quarkus/blob/master/devtools/common/src/main/filtered/extensions.json

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 24]
```diff
@@ -565,3 +568,483 @@ memory in a runtime class loader. To have these classes written out to disk for
...
+.Register a Bean Definining Annotation
+[source,java]
+----
+import javax.enterprise.context.ApplicationScoped;
+import org.jboss.jandex.DotName;
+import io.quarkus.extest.runtime.TestAnnotation;
+
+public final class TestProcessor {
+    static DotName TEST_ANNOTATION = DotName.createSimple(TestAnnotation.class.getName());
+    static DotName TEST_ANNOTATION_SCOPE = DotName.createSimple(ApplicationScoped.class.getName());
```
Comment (mkouba) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: You can replace this with `io.quarkus.arc.processor.BuiltinScope.APPLICATION.getInfo().getDotName()`. Maybe, we should add a convenient method like `BuiltinScope.getName()`. Also the defaultScope is not mandatory - `@Dependent` is used by default.

File: docs/src/main/asciidoc/extension-authors-guide.adoc (hunk 3 above)
Comment (mkouba) [reply docs/src/main/asciidoc/extension-authors-guide.adoc]: We should IMHO add TOC to all guides - recently, I created https://github.com/quarkusio/quarkus/issues/1490

**Comment by starksm64** [comment]
The CI build failures should not be resolved as the devtools dependency in the core/test-extension have been moved into the integration/test-extension test module

File: core/test-extension/runtime/pom.xml (hunk 1 above)
Comment (Sanne) [reply core/test-extension/runtime/pom.xml]: what's the resolution here?  There's lots of good stuff in this PR, I'd rather merge it than wait.

Is it important we have the quarkus-jaxb module if this jaxb dep is only used for testing? @gsmet could you open a separate issue if you think so? I'll merge soon if I don't see blockers :)

Comment (starksm64) [reply core/test-extension/runtime/pom.xml]: Right, I would rather get the content out and then fix the dependency issue for the next release once I can verify that it does not introduce the issues like the native image build dependency caused. I created an issue for this. #1819

File: docs/src/main/asciidoc/extension-authors-guide.adoc (hunk 7 above)
Comment (gsmet) [reply docs/src/main/asciidoc/extension-authors-guide.adoc]: @starksm64 could you address this one? The statement is false AFAIK. The feature is just there to be displayed in the feature list at startup.

**Comment by gsmet** [comment]
@starksm64 it looks like some of my comments were not addressed. Could you take a look, please?

File: docs/src/main/asciidoc/extension-authors-guide.adoc (hunk 7 above)
Comment (starksm64) [reply docs/src/main/asciidoc/extension-authors-guide.adoc]: Ok, right. I have corrected that section to indicate that the feature name should match a label in the extensions.json entry so that there is a correspondence between the displayed name and the label which is actually what is used to select the extension.

**Comment by starksm64** [comment]
Ok, not sure why I keep missing some of your comments, but I have addressed the majority of them.

**Comment by Sanne** [comment]
there's a failure with JDK11:
```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:build (default) on project quarkus-integration-test-elytron-security: Failed to build a runnable JAR: Failed to build a runner jar: Failed to augment application classes: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.elytron.security.deployment.SecurityDeploymentProcessor#registerJCAProviders threw an exception: java.lang.IllegalAccessError: class io.quarkus.elytron.security.deployment.SecurityDeploymentProcessor (in unnamed module @0xad66359) cannot access class sun.security.jca.Providers (in module java.base) because module java.base does not export sun.security.jca to unnamed module @0xad66359

```

**Comment by starksm64** [comment]
Strange, this seemed to pass previously. I have removed that import as it was unused.

**Comment by dmlloyd** [comment]
We've just recently become strict about using non-standard classes, in order to prevent build errors of the flavor previously seen when the project is partially built with Java 9 or later and then a native image is produced (unsuccessfully due to referencing Java 9+ symbols and overloads).

**Comment by starksm64** [comment]
@Sanne should be good to go

**Comment by Sanne** [comment]
thanks @starksm64 . I see you merged master into your feature branch, isn't that confusing? You mind if I rebase it one last time?

**Comment by starksm64** [comment]
No, that is fine. I'm not sure how that happened.

**Review by gsmet — APPROVED** [review]
Approved and tested by Sanne.

**Comment by Sanne** [comment]
rebased and added a micro update: 49f617404

I'll need @gsmet to merge it as I no longer can push the button. Thanks!
